### ISSUE TITLE
/copy method works synchronously

### ISSUE THREAD

**Issue by AndrewPashkin** [OneDrive/onedrive-api-docs#762]
I've noticed that `/copy` method works synchronously for me, here is the complete request-response dump:
```
< POST /_api/v2.0/drive/items/01E3ZKANTK7BRCRMNGBFHLTLQVDFHHNP7A/copy/ HTTP/1.1
< Host: pashkin-my.sharepoint.com
< User-Agent: python-requests/2.18.4
< Accept-Encoding: gzip, deflate
< Accept: */*
< Connection: keep-alive
< Authorization: Bearer <TOKEN>
< Content-Length: 125
< Content-Type: application/json
< 
< {"parentReference": {"driveId": "b!vg3hQ1ggjkGGOWPs0R9Du-E9zxMqD3ZKsSHqHEkMGjp1XrrKrEbnT6MXqyUH8dwu", "path": "test_folder"}}
> HTTP/1.1 202 Accepted
> Cache-Control: private, max-age=0
> Expires: Wed, 13 Dec 2017 22:33:01 GMT
> Last-Modified: Thu, 28 Dec 2017 22:33:01 GMT
> Location: https://pashkin-my.sharepoint.com/personal/pashkin_pashkin_onmicrosoft_com/_api/v2.0/drives/b!vg3hQ1ggjkGGOWPs0R9Du-E9zxMqD3ZKsSHqHEkMGjp1XrrKrEbnT6MXqyUH8dwu/items/01E3ZKANUV6WXNG5OWVZEKHKEZL4WVNF5S?force303=1&tempauth=<TEMP_AUTH_TOKEN>
> P3P: CP="ALL IND DSP COR ADM CONo CUR CUSo IVAo IVDo PSA PSD TAI TELo OUR SAMo CNT COM INT NAV ONL PHY PRE PUR UNI"
> X-SharePointHealthScore: 0
> X-SP-SERVERSTATE: ReadOnly=0
> ODATA-VERSION: 4.0
> SPClientServiceRequestDuration: 291
> SPRequestDuration: 382
> X-AspNet-Version: 4.0.30319
> SPRequestGuid: eb163b9e-402f-4000-ae81-ff9b46d22183
> request-id: eb163b9e-402f-4000-ae81-ff9b46d22183
> MS-CV: njsW6y9AAECugf+bRtIhgw.0
> Strict-Transport-Security: max-age=31536000
> X-FRAME-OPTIONS: SAMEORIGIN
> X-Powered-By: ASP.NET
> MicrosoftSharePointTeamServices: 16.0.0.7206
> X-Content-Type-Options: nosniff
> X-MS-InvokeApp: 1; RequireReadOnly
> X-MSEdge-Ref: Ref A: 9C648D5235AC490E9FFE4EE5C7E89247 Ref B: AMSEDGE0507 Ref C: 2017-12-28T22:33:01Z
> Date: Thu, 28 Dec 2017 22:33:00 GMT
> Content-Length: 0
> 
```

The [docs][copy] say that the method should something like that:
```
HTTP/1.1 202 Accepted
Location: https://contoso.sharepoint.com/_api/v2.0/monitor/4A3407B5-88FC-4504-8B21-0AABD3412717
```

But in my case I received:
```
HTTP/1.1 202 Accepted
Location: https://pashkin-my.sharepoint.com/personal/pashkin_pashkin_onmicrosoft_com/_api/v2.0/drives/b!vg3hQ1ggjkGGOWPs0R9Du-E9zxMqD3ZKsSHqHEkMGjp1XrrKrEbnT6MXqyUH8dwu/items/01E3ZKANUV6WXNG5OWVZEKHKEZL4WVNF5S?force303=1&tempauth=<TEMP_AUTH_TOKEN>
```
Notice that in my case I don't receive `/monitor` URL but instead I just receive URL of the copied item.

[copy]: https://docs.microsoft.com/en-us/onedrive/developer/rest-api/api/driveitem_copy

**Comment by ificator** [comment]
You're correct - currently the OneDrive for Business implementation of copy is a synchronous operation that pretends to be asynchronous so that it matches the calling pattern defined by OneDrive for Consumers. Fixing this is high on our list - we're hoping we'll have something to report in the next month or so.

**Comment by AndrewPashkin** [comment]
@ificator 
> copy is a synchronous operation that pretends to be asynchronous

But it doesn't return monitor-URL.

**Comment by ificator** [comment]
It returns a URL that should behave like a monitor URL does when the operation is complete, that is, if you hit it it'll 303 you to the newly created item. You won't get any progress as the copy is completed before the `/copy` request returns.

**Comment by AndrewPashkin** [comment]
@ificator, does it always return item-URL instead of monitor-URL or only for small files? Could it be clarified in the docs?

**Comment by ificator** [comment]
OneDrive for Business will behave like this for all files and folders. It's not called out in the documentation because what is returned does obey the documented contract for copy (assuming that you do actually see a 303 when you hit the URL returned from the copy request). If what you're seeing is the item metadata being returned immediately when you hit the URL returned from copy that would be a bug.

To make that a little more clear:

Expected=
1. Make copy request, get back a URL in the `Location` header
2. Make request to the URL from 1.
3. Get a 303 to another URL that looks pretty much exactly the same as the URL from 1 but with a different query string
4. Make a request to the URL from 3.
5. Get a 200 with the new item metadata

NOT Expected=
1. Make copy request, get back a URL in the `Location` header
2. Make request to the URL from 1.
3. Get a 200 with the new item metadata

**Comment by ificator** [comment]
Hi @AndrewPashkin - I apologize... I see what you're saying. In this case even though we're returning a 303 you're expecting the JSON body correct? It does indeed seem like we aren't complying with that aspect of the documentation.

**Comment by AndrewPashkin** [comment]
@ificator, the docs say<sup>([1][1], [2][2])</sup>:
> The value of the Location header provides a URL for a service that will return the current state of the copy operation. You can use this info to determine when the copy has finished.

and

> To check on the status of the copy action, the app makes a request to the URL provided in the previous response.

But instead of such URL (which I call monitor-URL) I get just item-URL.

And the documentation doesn't say that upon completion the item metadata should be returned, it [says][3] that something like that should be returned:
```
{
    "percentageComplete": 100.0,
    "resourceId": "01MOWKYVJML57KN2ANMBA3JZJS2MBGC7KM",
    "status": "completed"
}
```



[1]: https://github.com/OneDrive/onedrive-api-docs/blob/227cdc2136c6fc880a9a4f148d024850e7bd09aa/docs/rest-api/api/driveitem_copy.md
[2]: https://github.com/OneDrive/onedrive-api-docs/blob/227cdc2136c6fc880a9a4f148d024850e7bd09aa/docs/rest-api/concepts/long-running-actions.md#retrieve-a-status-report-from-the-monitor-url
[3]: https://github.com/OneDrive/onedrive-api-docs/blob/227cdc2136c6fc880a9a4f148d024850e7bd09aa/docs/rest-api/concepts/long-running-actions.md#retrieve-a-completed-status-report-from-the-monitor-url

**Comment by AndrewPashkin** [comment]
@ificator, actually I'm wrong. I was confused by the redirection, sorry for misinformation and thank you for assistance.
